### DISCUSSION TITLE
Event Based Scheduling

### DISCUSSION THREAD

**Discussion by Miksus** [Miksus/rocketry#123] [Ideas]
Hi!

I'm drafting ideas for event-based support for Rocketry. This would not shift the paradigm of Rocketry. Event-based scheduling fits quite natively inside statement-based system, I think.

The following are just preliminary ideas which might change (a lot), or be discarded. Please share your own whether you think this is a bad idea or a good one.

---

I'm thinking of supporting event factories which are functions that either return ``None`` (no event) or something else (event occurred).

It could look like this:

```python

@app.event(name="person")
def get_person():
    person = get_next_person_from_database()
    if person:
        # Event exists, a new person exists in the database
        return person
    else:
        # No event, task won't run
        return None

@app.task(get_person)
def process_person(person):
    ...

``` 

(The ``if`` block is actually unnecessary, I just over-emphasized it to highlight the mechanism.)

Basically these event are sort of streams/flows: they fetch an item and if an item is found, the task is started and the item is passed to the task. The ``name`` specifies the parameter key for which the event value is set to. 

---

Futhermore, the events are just conditions with some sort of parametrization logic. The could be used with other conditions as well:

```python
@app.event(name="person")
def get_person():
    person = get_next_person_from_database()
    return person

@app.task(time_of_day.between("10:00", "18:00") & get_person)
def process_person(person):
    ...
```

---

Remaining issues:

1. If multiple tasks use the same event, then how to make sure they all get executed and the first one won't consume the event.
2. Anything else?

Feel free to share your ideas.

**Comment by Miksus** [comment]
There is a dilemma: if an event occurs and the task is running, it won't get triggered (due to the run lock). The run lock could be optionally overridden in a future release but there is still a problem with the parameters: how to store them till the task starts? 

Maybe this is a place for the batch run (``task.run``) mechanism. It should be pretty easy to implement via that but the problem here is that it could no longer be arbitrarily combined with the conditions as the batch mechanism override the conditions. 

Or maybe it could be both: the event can be set as a condition but then there was also a separate trigger mechanism.
